modeshape-jcr/src/main/java/org/modeshape/jcr/locking/StandaloneLockingService.java

```diff
@@ -0,0 +1,123 @@
+        CheckArg.isNonNegative(lockTimeoutMillis, "lockTimeoutMillis");
+        this.lockTimeoutMillis = lockTimeoutMillis;
+    }
+
+    @Override
+    public boolean tryLock(long time, TimeUnit unit, String... names) {
+        Set<ReentrantLock> successfullyLocked = new HashSet<>();
+        for (String name : names) {
+            LOGGER.debug("Attempting to lock {0}", name);
+            ReentrantLock lock = null;
+            boolean success = false;
+            try {
+                lock = locksByName.computeIfAbsent(name, (lockName) -> new ReentrantLock());
+                success = time > 0 ? lock.tryLock(time, unit) : lock.tryLock();
+            } catch (InterruptedException e) {
+                LOGGER.debug("Thread {0} received interrupt request while waiting to acquire lock '{1}'", 
+                             Thread.currentThread().getName(), name);
+                Thread.currentThread().interrupt();
+            } catch (Exception e) {
+                LOGGER.debug(e, "Unexpected exception while attempting to lock '{0}'", name);
+            }
+            // if lock acquisition was not successful (for whatever reason) revert the lock
+            if (!success) {
+                LOGGER.debug("Unable to acquire lock on {0}. Reverting back the already obtained locks: {1}", name,
+                             successfullyLocked);
+                    successfullyLocked.stream().forEach(ReentrantLock::unlock);
+                return false;
+            }
+            LOGGER.debug("{0} locked successfully", name);
+            successfullyLocked.add(lock);
+        }
+        return true;
+    }
+
+    @Override
+    public List<String> unlock(String... names) {
+        return Arrays.stream(names)
+                     .map(this::unlock)
+                     .filter(Optional::isPresent)
+                     .map(Optional::get)
+                     .collect(Collectors.toList());
+    }
+    
+    private Optional<String> unlock(String name) {
+        LOGGER.debug("Attempting to unlock {0}", name);
+        ReentrantLock lock = locksByName.get(name);
+        if (lock == null) {
+            return Optional.of(name);
+        }
+        try {
+            lock.unlock();
+        } catch (IllegalMonitorStateException e) {
+            LOGGER.warn(JcrI18n.warnAnotherThreadAttemptingToUnlock, Thread.currentThread().getName(), lock);
+        }
+        if (lock.isLocked()) {
+            LOGGER.debug("Lock '{0}' is still locked with hold count '{1}'", name, lock.getHoldCount());
+            return Optional.of(name);
+        } else {
+            LOGGER.debug("'{0}' unlocked", name);
+            // if there is no contention for this lock, remove it
+            // otherwise leave it because the last queued thread will remove it once it completes
+            if (!lock.hasQueuedThreads()) {
+                locksByName.remove(name);
+            }
+            return Optional.empty();
```

Comment:
What happens if the lock is not locked on line 110, but by the time it gets here it has been locked again before we reach line 117? If there was only one other thread that acquired the lock, then the # of queued threads would be zero and we'd remove it from the map even though another thread is blocked on line 69 above attempting to lock the *existing* lock. (It would eventually unblock and acquire the lock, but that lock would no longer be in the map. That'd be bad.)

Also, it's not clear why you are returning a non-empty optional on line 112, since it might be possible that the lock was successfully unlocked for this thread by immediately acquired by another thread. So the return logic in this method seems to return a name when the named lock is still locked by somebody, even if it were successfully unlocked by me and I no longer hold any of the named locks. OTOH, the [JavaDoc for `LockingService.unlock(String...names)`](https://github.com/ModeShape/modeshape/pull/1496/files#diff-f872a09a1f66c182f534be128e5fa5aeR60) seems to imply that method returns only the names of those locks held by this thread that could not be released/unlocked -- essentially what of *my locks* could not be unlocked (which should only happen because unlock was called by a thread that did not acquire the lock).

How about the following logic, which I think fixes both problems?

    private Optional<String> unlock(String name) {
        LOGGER.debug("Attempting to unlock {0}", name);
        AtomicBoolean unlocked = new AtomicBoolean(false);
        locksByName.computeIfPresent(name, (key,lock)->{
            // Try to unlock the lock ...
            try {
                lock.unlock();
                unlocked.set(true);
            } catch (IllegalMonitorStateException e) {
                LOGGER.warn(JcrI18n.warnAnotherThreadAttemptingToUnlock, Thread.currentThread().getName(), lock);
                return lock;
            }
            if (lock.isLocked()) {
                LOGGER.debug("Lock '{0}' is still locked with hold count '{1}'", name, lock.getHoldCount());
                return lock;
            } else if ( lock.hasQueuedThreads() ) {
                LOGGER.debug("Lock '{0}' is not currently locked by will be", name);
                return lock;
            }
            // The lock is not used, and returning null will remove it from the map ...
            return null;
        });
        // A non-null reference to the lock means the lock is still there ...
        return unlocked.get() ? Optional.empty() : Optional.of(name);
    }

This does all of the work on the lock atomically with respect to the concurrent map, meaning it should not be possible for a race condition to occur between locking and unlocking a single lock. This logic is equivalent to your logic, except that it returns the name only when this thread's lock could not be released, which sounds like it better satisfies the JavaDoc intent.

What do you think?